### DISCUSSION TITLE
firebase_app_distribution failed

### DISCUSSION THREAD

**Discussion by usama-shafiq** [fastlane/fastlane#17436] [Q&A]
I created a fastfile which is used for app distribution:

`firebase_app_distribution(
        app: ENV["FIREBASE_APP_ID"],
        firebase_cli_token: ENV["FIREBASE_TOKEN"],
        groups: ""
    )`

It is giving error:

```
[09:42:33]: App Distribution could not generate credentials from the refresh token specified.

+------+------------------------------+-------------+
|                 fastlane summary                  |
+------+------------------------------+-------------+
| Step | Action                       | Time (in s) |
+------+------------------------------+-------------+
| 1    | Verifying fastlane version   | 0           |
| 2    | default_platform             | 0           |
| 3    | setup_circle_ci              | 0           |
| 4    | match                        | 1           |
| 5    | gym                          | 1131        |
| 6    | cd .. && touch changelog.txt | 0           |
| 7    | last_git_commit              | 0           |
| 💥   | firebase_app_distribution    | 0           |
+------+------------------------------+-------------+

[09:42:33]: fastlane finished with errors

[!] App Distribution could not generate credentials from the refresh token specified.
```

Any help will be appreciated.

**Comment by rogerluan** [comment]
Hey @usama-shafiq 👋 

Could you provide info from our [issue template](https://raw.githubusercontent.com/fastlane/fastlane/master/.github/ISSUE_TEMPLATE/BUG_REPORT.md)? 🙏

**Comment by techinpark** [comment]
@usama-shafiq  Are you installed `firebase cli`? 

Follow this step. 

1. Install firebase tools 
```sh 
$ curl -sL firebase.tools | bash 
```

2. Login firebase tool cli for ci 

```sh 
$ firebase login:ci 

In this step you can getting `FIREBASE_APP_TOKEN`
```

3. input your `FIREBASE_APP_TOKEN` in your `.env` file. 

4. try again 
```sh
firebase_app_distribution( app: ENV["FIREBASE_APP_ID"], firebase_cli_token: ENV["FIREBASE_TOKEN"], groups: "" )
```

**Reply by blixit** [reply]
using the `firebase` command helped me to re-auth. Thanks a lot

**Reply by milanvadhel-mi** [reply]
Hello @techinpark I have tried above steps and it has generated new token. I have added fresh new token in github secrets. I able to upload the build to firebase app distribution with same token via fastlane but problem is that when i try to upload the build with gihub action on github server then it shows me **App Distribution could not generate credentials from the refresh token specified.**. I have tried to reauth and generate new token but it won't work. Can you help me?

**Reply by techinpark** [reply]
> Hello @techinpark I have tried above steps and it has generated new token. I have added fresh new token in github secrets. I able to upload the build to firebase app distribution with same token via fastlane but problem is that when i try to upload the build with gihub action on github server then it shows me **App Distribution could not generate credentials from the refresh token specified.**. I have tried to reauth and generate new token but it won't work. Can you help me?

Can you share your action yml code?

**Reply by milanvadhel-mi** [reply]
@techinpark That issue is fixed now. I was accessing the FIREBASE_TOKEN in wrong way.

**Comment by jongwonatwork** [comment]
Hi, @usama-shafiq 
I have the same problem.
Have you solved the problem?

**Comment by sudachi808** [comment]
@usama-shafiq @jongwonatwork 

Use "Refresh Token".

```bash
$ fastlane run firebase_app_distribution_login
[✔] 🚀 
+-------------------------------------------+---------+----------------------------------+
|                                      Used plugins                                      |
+-------------------------------------------+---------+----------------------------------+
| Plugin                                    | Version | Action                           |
+-------------------------------------------+---------+----------------------------------+
| fastlane-plugin-firebase_app_distribution | 0.2.4   | firebase_app_distribution_login  |
|                                           |         | firebase_app_distribution        |
+-------------------------------------------+---------+----------------------------------+

[15:58:10]: ---------------------------------------------
[15:58:10]: --- Step: firebase_app_distribution_login ---
[15:58:10]: ---------------------------------------------
[15:58:10]: Open the following address in your browser and sign in with your Google account:
[15:58:10]: https://accounts.google.com/o/oauth2/auth?
[15:58:10]: 
[15:58:10]: Enter the resulting code here: *******************
[16:05:27]: 
[16:05:27]: Set the refresh token as the FIREBASE_TOKEN environment variable
[16:05:27]: Refresh Token: *******************                      <-- USE THIS!!
[16:05:27]: Result: true

```

**Reply by jongwonatwork** [reply]
@sudachi808  Thanks.
But it's not working for me.
``` shell
[09:05:41]: $ fastlane run firebase_app_distribution_login
[09:05:42]: ---------------------------------------------
[09:05:42]: --- Step: firebase_app_distribution_login ---
[09:05:42]: ---------------------------------------------
[09:05:42]: Open the following address in your browser and sign in with your Google account:
[09:05:42]: https://accounts.google.com/o/oauth2/auth?
[09:05:42]: 
[09:05:42]: Enter the resulting code here: 4/1AY0e-********
[09:06:15]: The code you entered is invalid. Copy and paste the code and try again.   <-- THIS IS MY PROBLEM :-)
[09:06:15]: Result: true
```

**Reply by Maria218** [reply]
> @usama-shafiq @jongwonatwork
> 
> Use "Refresh Token".
> 
> ```shell
> $ fastlane run firebase_app_distribution_login
> [✔] 🚀 
> +-------------------------------------------+---------+----------------------------------+
> |                                      Used plugins                                      |
> +-------------------------------------------+---------+----------------------------------+
> | Plugin                                    | Version | Action                           |
> +-------------------------------------------+---------+----------------------------------+
> | fastlane-plugin-firebase_app_distribution | 0.2.4   | firebase_app_distribution_login  |
> |                                           |         | firebase_app_distribution        |
> +-------------------------------------------+---------+----------------------------------+
> 
> [15:58:10]: ---------------------------------------------
> [15:58:10]: --- Step: firebase_app_distribution_login ---
> [15:58:10]: ---------------------------------------------
> [15:58:10]: Open the following address in your browser and sign in with your Google account:
> [15:58:10]: https://accounts.google.com/o/oauth2/auth?
> [15:58:10]: 
> [15:58:10]: Enter the resulting code here: *******************
> [16:05:27]: 
> [16:05:27]: Set the refresh token as the FIREBASE_TOKEN environment variable
> [16:05:27]: Refresh Token: *******************                      <-- USE THIS!!
> [16:05:27]: Result: true
> ```

I tried this but get the same error @jongwonatwork gets. No matter how many times I try, I get the error "The code you entered is invalid. Copy and paste the code and try again" instead of getting the refresh token. What could this mean?

**Comment by mnlightstone** [comment]
Hi all, I'm getting the same error as @Maria218 (`The code you entered is invalid. Copy and paste the code and try again"`). 

Did anyone figure out a resolution?
